File: Client/main/java/com/example/securechat/crypto/KeyExchangeFactory.java
package com.example.securechat.crypto;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class KeyExchangeFactory {
    private Map<String, Class<? extends KeyExchange>> algorithms;

    public KeyExchangeFactory() {
        algorithms = new HashMap<>();
        registerDefaultAlgorithms();
    }

    private void registerDefaultAlgorithms() {
        algorithms.put("ecdh", ECDHKeyExchange.class);
        algorithms.put("ecdh_2", ECDHKeyExchange2.class);
        algorithms.put("ecdh_3", ECDHKeyExchange2.class);
    }

    public void register(String name, Class<? extends KeyExchange> algorithmClass) {
        algorithms.put(name.toLowerCase(), algorithmClass);
    }

    public KeyExchange create(String name) throws Exception {
        Class<? extends KeyExchange> algorithmClass = algorithms.get(name.toLowerCase());
        if (algorithmClass == null) {
            throw new IllegalArgumentException("Unknown key exchange algorithm: " + name);
        }
        return algorithmClass.newInstance();
    }

    public Set<String> getSupportedAlgorithms() {
        return algorithms.keySet();
    }

    public boolean isSupported(String name) {
        return algorithms.containsKey(name.toLowerCase());
    }
}

File: Server/src/crypto/keyExchangeFactory.js
import { ECDH } from './ecdh.js';
import { ECDH_2 } from './ecdh_2.js';
import { ECDH_3 } from './ecdh_3.js';
import { KeyExchange } from './keyExchange.js';
// Import future implementations here
// import { RSA } from './rsa.js';
// import { DH } from './dh.js';

export class KeyExchangeFactory {
  constructor() {
    this.algorithms = new Map();
    this.registerDefaultAlgorithms();
  }

  // Register default algorithms
  registerDefaultAlgorithms() {
    this.register('ecdh', ECDH);
    this.register('ecdh_2', ECDH_2);
    this.register('ecdh_3', ECDH_3);
  }

  // Register a new key exchange algorithm
  register(name, algorithmClass) {
    if (!(algorithmClass.prototype instanceof KeyExchange) && algorithmClass !== KeyExchange) {
      throw new Error(`${algorithmClass.name} must extend KeyExchange`);
    }
    this.algorithms.set(name.toLowerCase(), algorithmClass);
  }

  // Create instance of key exchange algorithm with optional parameters
  create(name, params = null) {
    const AlgorithmClass = this.algorithms.get(name.toLowerCase());
    if (!AlgorithmClass) {
      throw new Error(`Unknown key exchange algorithm: ${name}`);
    }
    
    // Pass parameters to constructor for algorithms that support it
    if (name.toLowerCase() === 'ecdh' && params) {
      return new AlgorithmClass(params);
    } else if (name.toLowerCase() === 'ecdh_2' && params) {
      return new AlgorithmClass(params);
    } else if (name.toLowerCase() === 'ecdh_3' && params) {
      return new AlgorithmClass(params);
    }
    
    return new AlgorithmClass();
  }

  // Get list of supported algorithms
  getSupportedAlgorithms() {
    return Array.from(this.algorithms.keys());
  }

  // Check if algorithm is supported
  isSupported(name) {
    return this.algorithms.has(name.toLowerCase());
  }
}